### ISSUE TITLE
Setup Error

### ISSUE THREAD

**Issue by en2te** [lociii/homeassistant-digitalstrom#1]
Thanks for doing the HACS-Integration! I am really looking forward to getting this up and running. I installed the Integration but when i do "integrations" => "+" i get the message "Unknown error occured". This is the detailled error message:

```
Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_protocol.py", line 418, in start
    resp = await task
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_middlewares.py", line 119, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/real_ip.py", line 37, in real_ip_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 69, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 228, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 124, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 162, in post
    return await super().post(request, flow_id)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 47, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 90, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 96, in async_configure
    result = await self._async_handle_step(flow, cur_step["step_id"], user_input)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 139, in _async_handle_step
    result = await getattr(flow, method)(user_input)  # type: Dict
  File "/config/custom_components/digitalstrom/config_flow.py", line 67, in async_step_user
    if self.device_config[CONF_SLUG] in configured_devices(self.hass):
KeyError: 'slug'
```

"digitalSTROM" appears in the "discovered" - Section but when i click "configure" i get an +
 popup with an endless spinning wheel and nothing else.

**Comment by lociii** [comment]
Hey, thanks for the report.
This whole config flow is a mess. I will need to rework it soon.
However I got it installed with the current commit.

**Comment by en2te** [comment]
Uninstall / Install of the component did not help. May be some relicts from earlier manual installation attempts on my system? Tried it about half a year ago with pydigitalstrom and your component but never got it running.

**Comment by lociii** [comment]
Hey there. Would you mind trying again.

**Cleanup**

Please make sure all existing integrations are removed.
Maybe also check for a subdirectory .digitalstrom in your HA config folder (this is where the python lib stores the app token), remove if exists.
Maybe also check .storage/core.config_entries in your HA config folder for any digitalstrom related entries and remove them while HA is not running.

**Comment by en2te** [comment]
Yes, it is working! Had to use the IP instead of dss.local. Thanks for your work! I see only "entities without devices". Can you give me a hint how to get access to the single devices?

**Comment by lociii** [comment]
I've just updated the readme on how this integration works with digitalSTROM.
TL;DR no device is exposed directly, only scenes are controlled.

see https://github.com/lociii/homeassistant-digitalstrom#what-does-it-do

**Comment by en2te** [comment]
But scenes are stateless, right? Can you see if a specific light is on or a window is open? How do you do this? Switches with scripts that activates the on / off scene? Would be nice if you could add an example to the readme. Last thing: Some Scenes are not exposed to HA and i can't see any difference in dssadmin. Is it possible to force an update?

**Comment by lociii** [comment]
> But scenes are stateless, right? Can you see if a specific light is on or a window is open? How do you do this? Switches with scripts that activates the on / off scene? 
Exactly. The light switch for area 1 is calling light scene 6 to turn the light(s) on and scene 1 to turn them off.
To keep HA in sync, the integration registers a listener that pulls DS updates every 1 second. Scene calls are reported to this listener and are forwarded to the devices to change their status.
There is also an undocumented websocket on the DS server which I will investigate soon. Maybe it can be used to subscribe to updates.
Does this answer your question?

> Would be nice if you could add an example to the readme. 
Not yet sure what kind of example you would like to see. Feel free to improve the readme and open a pull request.

> Last thing: Some Scenes are not exposed to HA and i can't see any difference in dssadmin. Is it possible to force an update?
Do they have a name?
The [underlying library](https://github.com/lociii/pydigitalstrom#concept) only supports named scenes for now. Mainly because I'm too lazy to think about naming conventions when I have to generate names :)

**Comment by en2te** [comment]
Switching Scenes with input_boolean works fine. on=>scene-on off=>scene-off. The only thing i miss is the status-update if the scene was switched by the wall-switch or dssadmin Timer-App ... but to be honest: i don't use the wall-switch often and automations will be moved to Home-Assistant. 

There are scenes in dssadmin that i can not find in Home-Assistant. Maybe they had no name other than "Stimmung x" or "Auf" / "Zu" when i installed your integration but now they have a name and don't show up yet ... but i can see scenes in HA that had names before the integration.

Another thing that you should probably fix: Put a "buy me a coffee" - Button somewhere ;-).

**Comment by lociii** [comment]
If you're familiar with python you could build something similar to the switch part of the integration. It combines the on and off scenes for presence and sleeping into a switch. Since all scene calls are propagated to the component, it can decide if it wants to update it's state based on the call 
Unfortunately there is no generic way of doing it since I have no idea which scene in a digitalSTROM system is the counterpart of another scene.
This completely depends on your setup 

For missing scenes, just restart HA and it should discover the missing scenes.

If you want to, we can also have a call to exchange ideas. Let me know

**Comment by en2te** [comment]
I updated to 1.1.0. The entities were found but when i restart HA all entities are not availlable. Deleting the integration and reinstalling (including the .digitalstrom) does the same. All entities accessible until i restart. The previous version worked fine.

Sorry, i am not familiar with python. This is the log:

`Error setting up entry Apartment (192.168.178.189:8080) for digitalstrom
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 924, in _wrap_create_connection
    await self._loop.create_connection(*args, **kwargs))
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 981, in create_connection
    ssl_handshake_timeout=ssl_handshake_timeout)
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 1009, in _create_connection_transport
    await waiter
ConnectionAbortedError: SSL handshake is taking longer than 60.0 seconds: aborting the connection

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/requesthandler.py", line 28, in raw_request
    async with session.get(url=url, **kwargs) as response:
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client.py", line 1005, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client.py", line 476, in _request
    timeout=real_timeout
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 522, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 854, in _create_connection
    req, traces, timeout)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 992, in _create_direct_connection
    raise last_exc
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 974, in _create_direct_connection
    req=req, client_error=client_error)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 931, in _wrap_create_connection
    raise client_error(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host 192.168.178.189:8080 ssl:None [None]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/digitalstrom/__init__.py", line 73, in async_setup_entry
    await client.initialize()
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/client.py", line 69, in initialize
    data = await self.request(url=self.URL_SCENES)
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/client.py", line 44, in request
    self._session_token = await self.get_session_token()
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/client.py", line 59, in get_session_token
    self.URL_SESSIONTOKEN.format(apptoken=self._apptoken)
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/requesthandler.py", line 41, in raw_request
    raise DSRequestException("request failed")
pydigitalstrom.exceptions.DSRequestException: request failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 190, in async_setup
    hass, self
  File "/config/custom_components/digitalstrom/__init__.py", line 75, in async_setup_entry
    raise PlatformNotReady("Failed to initialize digitalSTROM server at %s", client.host)
homeassistant.exceptions.PlatformNotReady: ('Failed to initialize digitalSTROM server at %s', 'https://192.168.178.189:8080')`

**Comment by lociii** [comment]
Thanks for the report. Please just use version 1.0.0 for now - still available in HACS.
I'll investigate the issue on the weekend. I've seen this behavior during my tests but is was not happening consistently.

**Comment by lociii** [comment]
2.1.2 has been released which works fine for me so far.
Major improvements:
* Use websocket instead of polling every second
* Use a command stack to delay execution (too fast commands kill the DS server)
* Store token in integration config instead of yaml

However there are a few non-critical things left over to be fixed:
* Websocket reconnect
* Websocket proper shutdown on restart of HA
* Initializing devices when setting up the integration (currently only on HA start)

**Comment by en2te** [comment]
Thanks for the new version - but i can't get it up and running.
Problem 1: In HACS i cannot upgrade. The installed version remains 1.0.0. I uploaded the custom_component - Folder per SFTP but the version number (shown) remains 1.0.0. Other components can be upgraded.
Problem 2: I removed the integration, removed the .digitalstrom - folder and hit the (+) in integratios. Filled in password and Server-Address but "unknown error" is all i get. 

This is from the error log. Tell me if i can help debugging. 

`
Protokolldetails ( ERROR )
Tue Nov 26 2019 00:42:25 GMT+0100 (Mitteleuropäische Normalzeit)
Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_protocol.py", line 418, in start
    resp = await task
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_app.py", line 458, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/web_middlewares.py", line 119, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/real_ip.py", line 40, in real_ip_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 73, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 136, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 122, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 163, in post
    return await super().post(request, flow_id)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 48, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 90, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 89, in async_configure
    result = await self._async_handle_step(flow, cur_step["step_id"], user_input)
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 132, in _async_handle_step
    result: Dict = await getattr(flow, method)(user_input)
  File "/config/custom_components/digitalstrom/config_flow.py", line 98, in async_step_user
    apartment_name=self.device_config[CONF_ALIAS],
TypeError: __init__() got an unexpected keyword argument 'username'
`

**Comment by lociii** [comment]
Do you use other HACS installed components? Do they update?
What's always confusing me in HACS is that it's not updating when it shows something like "Available (newest: 2.1.2)" but only when you click on the dropdown and then really select "2.1.2" and update.

Manually overwriting files will never update the version number since HACS has an internal database of revisions. They're not stored on the files of the component.

--

About the error: it seems like you're still running an outdated version of the component.
The last revision that contained the word "apartment_name" in line 98 of config_flow.py was 9c58853

**Comment by en2te** [comment]
Selecting the 2.1.2 by clicking did the trick. The latest version is now installed. 

I could add the integration (i had to use the IP-Address, dss.local don’t work) but there are no devices. After Home-Asssitant restart if found this In the log:


```
Protokolldetails ( ERROR )
Tue Nov 26 2019 11:53:18 GMT+0100 (Mitteleuropäische Normalzeit)
Error doing job: Task exception was never retrieved
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 936, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 981, in create_connection
    ssl_handshake_timeout=ssl_handshake_timeout)
  File "/usr/local/lib/python3.7/asyncio/base_events.py", line 1009, in _create_connection_transport
    await waiter
ConnectionAbortedError: SSL handshake is taking longer than 60.0 seconds: aborting the connection

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/requesthandler.py", line 29, in raw_request
    async with session.get(url=url, **kwargs) as response:
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client.py", line 1012, in __aenter__
    self._resp = await self._coro
  File "/usr/local/lib/python3.7/site-packages/aiohttp/client.py", line 483, in _request
    timeout=real_timeout
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 523, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 859, in _create_connection
    req, traces, timeout)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 1004, in _create_direct_connection
    raise last_exc
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 986, in _create_direct_connection
    req=req, client_error=client_error)
  File "/usr/local/lib/python3.7/site-packages/aiohttp/connector.py", line 943, in _wrap_create_connection
    raise client_error(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host 192.168.178.189:8080 ssl:None [None]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/digitalstrom/__init__.py", line 91, in digitalstrom_discover_devices
    await client.initialize()
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/client.py", line 77, in initialize
    response = await self.request(url=self.URL_SCENES)
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/client.py", line 56, in request
    self._session_token = await self.get_session_token()
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/client.py", line 67, in get_session_token
    self.URL_SESSIONTOKEN.format(apptoken=self._apptoken)
  File "/usr/local/lib/python3.7/site-packages/pydigitalstrom/requesthandler.py", line 42, in raw_request
    raise DSRequestException("request failed")
pydigitalstrom.exceptions.DSRequestException: request failed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/config/custom_components/digitalstrom/__init__.py", line 94, in digitalstrom_discover_devices
    "Failed to initialize digitalSTROM server at %s", client.host
homeassistant.exceptions.ConfigEntryNotReady: ('Failed to initialize digitalSTROM server at %s', '192.168.178.189')

```

**Comment by lociii** [comment]
Well. That's nasty.
Thought the workaround with delaying connect until HA is ready will solve it.
Thanks for the report. Seems that more research is necessary.

**Comment by en2te** [comment]
Now it works :). I just updated Home-Assistant. After the restart the DS Entities were there again and work. Very nice! Thanks for your work (again!)
